src/daft-dsl/src/join.rs

```diff
@@ -0,0 +1,99 @@
+use common_error::DaftResult;
+use daft_core::{prelude::*, utils::supertype::try_get_supertype};
+use indexmap::IndexSet;
+
+use crate::{deduplicate_expr_names, ExprRef};
+
+pub fn get_common_join_cols<'a>(
```

Comment:
Common columns now determined via schema instead of join keys so that join keys can be modified (ex: casting to supertype) without side effects. Does introduce a small API change in the order of the join schema: common columns are now sorted by left side schema instead of by join keys. I think this is fine but we could introduce a project under the left side to reorder if necessary.